### ISSUE TITLE
ContainerSet Retry strategy does nothing?

### ISSUE THREAD

**Issue by nicolas-vivot** [argoproj/argo-workflows#11502]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

Hi,

According to [this closed issue](https://github.com/argoproj/argo-workflows/issues/7290) we are supposed to be able to configure a retry strategy that applies to containers in a `ContainerSet`, not to the whole `Pod`.

After reading the documentation, i tried the following two configuration:

1. specifying a `retryStategy` with a `limit` of `2` at the template level (same level as `volumes`, `containerset`, `metadata`, etc.)
2. specifying a`retryStategy` with a `retries` of `2` at the containerset level (as specified in the above mentioned issue)

In case 1, the retry strategy applies to the entire `Pod`, meaning that the pod is restarted as any other container templates. The consequence is that all containers within this containerset are restarted, not just the container that failed.
![image](https://github.com/argoproj/argo-workflows/assets/79290619/f91d82d6-1a8e-4804-8f59-eeac59803038)

In case 2, no retry strategy is happening.
![image](https://github.com/argoproj/argo-workflows/assets/79290619/279ee4f9-a6c7-4cf1-8f2d-29f7df484b6e)

So, while the case 1 seems to work as expected (retrying all containers), the case 2 seems to retry nothing.
Checking logs, a single error log is logged, which probably validate the fact that no retry is happening.

Questions:

- do you really support restarting / retry strategy on a per container basis for `containerset` ?
- If yes, could you please indicate some clear documentation or workflow example that illustrate it ?

### Version

3.4.8

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: bug
  annotations:
    workflows.argoproj.io/description: |
      This workflow template is used to reproduce a bug.
spec:
  entrypoint: test
  templates:

    - name: test
      containerSet:
        retryStrategy:
          retries: "2"
        containers:
          - name: c1
            image: python:alpine3.6
            command:
              - python
              - -c
            args:
              - |
                print("hi")
          - name: c2
            image: python:alpine3.6
            command:
              - invalid
              - command
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

**Comment by terrytangyuan** [comment]
Yes it should be supported so this looks like a bug or a known limitation discussed in #7290. cc @alexec and @NikeNano

**Comment by alexec** [comment]
Not supported AFAIK. Retry is a pod level, not container level.

**Comment by nicolas-vivot** [comment]
> Not supported AFAIK. Retry is a pod level, not container level.

Thank you for your confirmation.
In such case, i do not really understand why we have the ability to setup a retry strategy at the `containerset` level, if it have exactly the same behavior as the retry strategy that we can set at the step level.

What was this [discussion](https://github.com/argoproj/argo-workflows/issues/7290) for, and these [adds](https://github.com/argoproj/argo-workflows/pull/7377/files) for ?

It confuses people without bringing anything useful, it should be removed then. Is my understanding correct ?

**Comment by NikeNano** [comment]
> @alexec Thank you for your confirmation.
> 
> In such case, what is the [retry strategy](https://github.com/argoproj/argo-workflows/commit/808c561f1c4a56668c32caa69be5b0441d372610) at the `containerset` for ?
> 
> At best, it confuses people since it does exactly the same as the retry strategy we can setup at the `step` level, but with different naming conventions and actually less possibilities.
> 
> At worse (i believe the real situation), it actually does nothing. No retry at all.
> 
> What was this [discussion](https://github.com/argoproj/argo-workflows/issues/7290) for, and these [adds](https://github.com/argoproj/argo-workflows/pull/7377/files) for ?
> 
> If it does not work, or do not bring anything, shouldn't it be removed then ?

I have sadly not have the time to work on Argo the last year(but hope to come back in the future) as I remember(and after reading through the conversations) this should be done on container level, where the goal is to rerun specific containers that failed. However it seems based upon your report this is not the case.

**Comment by shuangkun** [comment]
I find the root cause is we checked the command before retry.  https://github.com/argoproj/argo-workflows/blob/main/cmd/argoexec/commands/emissary.go#L114  So looks like do nothing, Not sure it can name a bug.

**Comment by shuangkun** [comment]
```
          - name: c2
            image: alpine:latest
            command: [sh, -c]
            args: ["echo intentional failure; exit 1"]
```
If change c2 to a good command, we can see it retrys
```
@B-QGHAMD6M-2124 11502 % kubectl logs bug-bgzfz -c c3
time="2024-03-15T06:21:08.933Z" level=info msg="capturing logs" argo=true
intentional failure
time="2024-03-15T06:21:08.951Z" level=debug msg="ignore signal child exited" argo=true
time="2024-03-15T06:21:09.940Z" level=info msg="capturing logs" argo=true
intentional failure
time="2024-03-15T06:21:09.943Z" level=debug msg="ignore signal child exited" argo=true
time="2024-03-15T06:21:10.944Z" level=info msg="sub-process exited" argo=true error="<nil>"
time="2024-03-15T06:21:10.945Z" level=info msg="not saving outputs - not main container" argo=true
Error: exit status 1
```

**Comment by agilgur5** [comment]
In reviewing https://github.com/argoproj/argo-workflows/pull/12809#pullrequestreview-1959420732, I also experienced some of this confusion. Hopefully my docs suggestions will clarify quite a bit, but as I wrote there, we should probably consider redoing the `ContainerSetRetryStrategy` spec entirely to match the `RetryStrategy` spec of other template types (which roughly wrap the [k8s wait package](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait), particularly for `backoff`).

I changed this issue to a "docs feature" since it is working as intended, it's just not documented clearly (or much at all).
